Python/suggestions.c

```diff
@@ -0,0 +1,230 @@
+                      PyObject* name,
+                      PyObject* oldexceptionvalue)
+{
+    assert(PyList_CheckExact(dir));
+
+    Py_ssize_t dir_size = PyList_GET_SIZE(dir);
+    if (dir_size >= MAX_GETATTR_PREDICT_ITEMS) {
+        return NULL;
+    }
+
+    int suggestion_distance = PyUnicode_GetLength(name);
+    PyObject* suggestion = NULL;
+    for (int i = 0; i < dir_size; ++i) {
+        PyObject *item = PyList_GET_ITEM(dir, i);
+        const char *name_str = PyUnicode_AsUTF8(name);
+        if (name_str == NULL) {
+            PyErr_Clear();
+            continue;
+        }
+        int current_distance = distance(PyUnicode_AsUTF8(name),
+                                        PyUnicode_AsUTF8(item));
+        if (current_distance > MAX_GETATTR_PREDICT_DIST){
+            continue;
+        }
+        if (!suggestion || current_distance < suggestion_distance) {
+            suggestion = item;
+            suggestion_distance = current_distance;
+        }       
+    }
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("%S\n\nDid you mean: %U?",
+                                oldexceptionvalue, suggestion);
+}
+
+int
+_Py_offer_suggestions_for_attribute_error(PyAttributeErrorObject* exc) {
+    int return_val = 0;
+
+    PyObject* name = exc->name;
+    PyObject* v = exc->obj;
+
+    // Aboirt if we don't have an attribute name or we have an invalid one
+    if ((name == NULL) || (v == NULL) || !PyUnicode_CheckExact(name)) {
+        return -1;
+    }
+    
+    PyObject* oldexceptionvalue = NULL;
+    Py_ssize_t nargs = PyTuple_GET_SIZE(exc->args);
+    switch (nargs) {
+        case 0:
+            oldexceptionvalue = PyUnicode_New(0, 0);
+            break;
+        case 1:
+            oldexceptionvalue = PyTuple_GET_ITEM(exc->args, 0);
+            Py_INCREF(oldexceptionvalue);
+            // Check that the the message is an uncode objects that we can use.
+            if (!PyUnicode_CheckExact(oldexceptionvalue)) {
+                return_val = -1;
+                goto exit;
+            }
+            break;
+        default:
+            // Exceptions with more than 1 value in args are not
+            // formatted using args, so no need to make a new suggestion.
+            return 0;
+    }
+
+    PyObject* dir = PyObject_Dir(v);
+    if (!dir) {
+        goto exit;
+    }
+
+    PyObject* newexceptionvalue = calculate_suggestions(dir, name, oldexceptionvalue);
+    Py_DECREF(dir);
+
+    if (newexceptionvalue == NULL) {
+        // We did't find a suggestion :(
+        goto exit;
+    }
+
+    PyObject* old_args = exc->args;
+    PyObject* new_args = PyTuple_Pack(1, newexceptionvalue);
+    Py_DECREF(newexceptionvalue);
+    if (new_args == NULL) {
+        return_val = -1;
+        goto exit;
+    }
+    exc->args = new_args;
+    Py_DECREF(old_args);
+
+exit:
+    Py_DECREF(oldexceptionvalue);
+
+    // Clear any error that may have happened.
+    PyErr_Clear();
```

Comment:
If the function clears unconditonnaly the current exception, I like to add an assertion to ensure that it's not called with an exception raised.